File: 2020/day-9/part_2.py
with open("input.txt", "r") as f:
    numbers = [int(n) for n in f.read().splitlines()]

invalid_number = 31161678  # this matches my input specifically

left, right = 0, 0
while True:
    subset = numbers[left : right + 1]
    total = sum(subset)
    if total == invalid_number:
        print(min(subset) + max(subset))
        break
    elif total < invalid_number:
        # we need more numbers
        right += 1
    else:  # total > invalid_number
        # we need less numbers
        left += 1

File: 2024/aoc2024/apps/day_01/part_2.exs
{ left, right } = Main.get_parsed("input.txt")

initial_counts = Enum.reduce(
  left,
  %{},
  fn val, acc -> Map.put(acc, val,
    case Map.get(acc, val) do
      nil -> {1, 0}
      { left, _ } -> {left + 1, 0}
    end)
end)

right
|> Enum.reduce(initial_counts, fn val, acc ->
    case Map.get(acc, val) do
      nil -> acc
      { left, right } -> Map.put(acc, val, {left, right + 1 })
    end
end)
|> Enum.map(fn {k, {left, right}} -> k*left*right end)
|> Enum.reduce(&(&1 + &2))
|> IO.puts()
